istioctl/pkg/multicluster/remote_secret.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2019 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/runtime/serializer/versioning"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth" // to avoid 'No Auth Provider found for name "gcp"'
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/tools/clientcmd/api/latest"
+
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/secretcontroller"
+)
+
+var (
+	Codec  runtime.Codec
+	Scheme *runtime.Scheme
+)
+
+func init() {
+	Scheme = runtime.NewScheme()
+	utilruntime.Must(v1.AddToScheme(Scheme))
+	yamlSerializer := json.NewYAMLSerializer(json.DefaultMetaFactory, Scheme, Scheme)
+	Codec = versioning.NewDefaultingCodecForScheme(
+		Scheme,
+		yamlSerializer,
+		yamlSerializer,
+		v1.SchemeGroupVersion,
+		runtime.InternalGroupVersioner,
+	)
+}
+
+type Options struct {
+	ServiceAccountName string
+	Contexts           []string
+
+	// inherited from root command
+	Namespace  string
+	Kubeconfig string
+}
+
+var (
+	defaultSecretPrefix       = "istio-pilot-remote-secret-"
+	DefaultServiceAccountName = "istio-pilot-service-account"
+)
+
+// NewCreatePilotRemoteSecretCommand creates a new command for joining two clusters togeather in a multi-cluster mesh.
+func NewCreatePilotRemoteSecretCommand(kubeconfig, namespace *string) *cobra.Command {
+	o := Options{
+		ServiceAccountName: DefaultServiceAccountName,
+	}
+
+	c := &cobra.Command{
+		Use:   "create-pilot-remote-secrets <cluster_context>... [OPTIONS]",
```

Comment:
How about 'create-remote-secret' ? First, this might very well move into galley ( so it can monitor remote clusters without having to run there), might also be moved into an extended citadel (same thing). And it doesn't seem to depend in any way on pilot. Operator too.

I would even go one step further and default to creating an 'istio-remote.istio-system' SA. 

This utility is likely more general than just Istio.

Reply:
`create-remote-secret` makes sense. 

> I would even go one step further and default to creating an 'istio-remote.istio-system' SA.

This new SA could also be added to the default charts or performed with kubectl (see below). Do you think there are cases where the user would want to use an existing SA and *not* create a new SA as part of generating the secret?

```bash
kubectl -n istio-system create serviceaccount istio-remote
kubectl create clusterrole istio-remote-istio-system --verb=get,list,watch --resource=pods,nodes,endpoints,services
kubectl create clusterrolebinding istio-remote-istio-system --clusterrole=istio-remote-istio-system --user=istio-system:istio-remote
```